### PR TITLE
[RFC][vm_runtime] Implement a substitution map in runtime to support generics

### PR DESCRIPTION
**Pull request by runtian-zhou**
## Motivation
In general the VM doesn't need to know about generic type parameters at runtime. The only exceptions are these global instructions (borrowGlobal, MoveFrom, MoveTo, Exists) as we need to know the exact struct layout to store them on chain. The current plan is that we are going to disallow top level struct to contain any type parameters to simplify the runtime implementation. i.e: We are going to store:
`Resource FooCoin{
    f: Foo<Coin>,
}`
not
`Resource FooCoin<T> {
    f: Foo<T>
}`
on chain.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Added a unit test. Will wait for the front end support for generics to perform the functional tests

### CODE REVIEW

File: language/vm/vm_runtime/src/substitution_map.rs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    /// Strings, immutable, utf8 representation.
+    String,
+    /// ByteArray, variable size, immutable byte array.
+    ByteArray,
+    /// Address, a 32 bytes immutable type.
+    Address,
+    /// MOVE user type, resource or unrestricted
+    Struct(StructHandleIndex, Vec<SignatureTokenRef<'a>>),
+    /// Type parameter.
+    TypeParameter(&'a SignatureTokenRef<'a>),
```
Comment (tnowacki) [inline language/vm/vm_runtime/src/substitution_map.rs]: This seems overly complicated, why not just use `SignatureToken`?

Comment (runtian-zhou) [reply language/vm/vm_runtime/src/substitution_map.rs]: I was trying to make copies only when necessary(when we try to materialize it)

Comment (tnowacki) [reply language/vm/vm_runtime/src/substitution_map.rs]: But why? I feel like you are going to need to copy this almost immediately.

Additionally, this adds an unusual layer of complexity, before we have any performance information

Comment (dariorussi) [reply language/vm/vm_runtime/src/substitution_map.rs]: I would agree with @tnowacki here that the effort seems too much.
I would start with using just SignatureToken and then see where it goes

**Comment by dariorussi** [comment]
overall I am not too sure what this is intended to do and the correlation between the PR and the diff. I will talk to you to understand this better...

**Comment by runtian-zhou** [comment]
As suggested by @tnowacki, we don't really need to use the reference to optimize for the performance for now.

**Comment by runtian-zhou** [comment]
This diff will cause bugs in struct definition cache. Will need to revise the design there.

**Pull request closed without merging**